### PR TITLE
UCS/MALLOC/FREE: use UCS allocators

### PR DESCRIPTION
**Pull request by hoopoepg**
- use UCS allocators for parser and client/server stats
- fixed potential ucs/malloc/free calls in error handlers

### CODE REVIEW

File: src/ucs/config/parser.c [hunk 1]
```diff
@@ -366,7 +366,7 @@ ucs_status_t ucs_config_clone_log_comp(const void *src, void *dst, const void *a
 
 int ucs_config_sscanf_bitmap(const char *buf, void *dest, const void *arg)
 {
-    char *str = strdup(buf);
+    char *str = ucs_strdup(buf, "config_sscanf_bitmap str");
```
Comment (yosefe) [inline src/ucs/config/parser.c:369]: can we use ucs_alloca() here?

File: src/ucs/config/parser.c [hunk 2]
```diff
@@ -693,7 +693,7 @@ int ucs_config_sscanf_range_spec(const char *buf, void *dest, const void *arg)
     char *p, *str;
     int ret = 1;
 
-    str = strdup(buf);
+    str = ucs_strdup(buf, "config_sscanf_range_spec str");
```
Comment (yosefe) [inline src/ucs/config/parser.c:696]: maybe add ucs_strdupa() macro to use ucs_alloca + memcpy() ?

File: src/ucs/config/parser.c [hunk 3]
```diff
@@ -1040,7 +1040,7 @@ static void ucs_config_print_doc_line_by_line(const ucs_config_field_t *field,
     char *doc, *line, *p;
     int num;
 
-    line = doc = strdup(field->doc);
+    line = doc = ucs_strdup(field->doc, "config_print_doc_line_by_line doc");
```
Comment (yosefe) [inline src/ucs/config/parser.c:1043]: "config_doc"

File: src/ucs/stats/client_server.c [hunk 4]
```diff
@@ -98,7 +99,7 @@ ucs_status_t ucs_stats_client_init(const char *server_addr, int port, ucs_stats_
     ucs_status_t status;
     int ret;
 
-    client = malloc(sizeof *client);
+    client = ucs_malloc(sizeof(*client), "stats_client_init");
```
Comment (yosefe) [inline src/ucs/stats/client_server.c:102]: "stats_client"

File: src/ucs/stats/client_server.c [hunk 5]
```diff
@@ -235,10 +236,12 @@ static void ucs_stats_server_entity_reset_buffer(stats_entity_t * entity,
     if (new_size != entity->buffer_size) {
         pthread_mutex_lock(&entity->lock);
         entity->buffer_size = new_size;
-        entity->inprogress_buffer = realloc(entity->inprogress_buffer,
-                                            new_size + sizeof(frag_hole_t));
-        entity->completed_buffer  = realloc(entity->completed_buffer,
-                                            new_size + sizeof(frag_hole_t));
+        entity->inprogress_buffer = ucs_realloc(entity->inprogress_buffer,
+                                                new_size + sizeof(frag_hole_t),
+                                                "stats entity inprogress_buffer");
```
Comment (yosefe) [inline src/ucs/stats/client_server.c:241]: "stats_inprogress_buffer"

File: src/ucs/stats/client_server.c [hunk 6]
```diff
@@ -235,10 +236,12 @@ static void ucs_stats_server_entity_reset_buffer(stats_entity_t * entity,
...
+        entity->completed_buffer  = ucs_realloc(entity->completed_buffer,
+                                                new_size + sizeof(frag_hole_t),
+                                                "stats entity completed_buffer");
```
Comment (yosefe) [inline src/ucs/stats/client_server.c:244]: "stats_completed_buffer"

File: src/ucs/stats/client_server.c [hunk 7]
```diff
@@ -560,7 +563,7 @@ ucs_status_t ucs_stats_server_start(int port, ucs_stats_server_h *p_server)
     ucs_stats_server_h server;
     ucs_status_t status;
 
-    server = malloc(sizeof *server);
+    server = ucs_malloc(sizeof *server, "stats_server_start");
```
Comment (yosefe) [inline src/ucs/stats/client_server.c:566]: "stats_server"

File: src/uct/rocm/ipc/rocm_ipc_cache.c [hunk 8]
```diff
@@ -217,7 +217,7 @@ ucs_status_t uct_rocm_ipc_create_cache(uct_rocm_ipc_cache_t **cache,
         goto err_destroy_rwlock;
     }
 
-    cache_desc->name = strdup(name);
+    cache_desc->name = ucs_strdup(name, "rocm cache->name");
```
Comment (yosefe) [inline src/uct/rocm/ipc/rocm_ipc_cache.c:220]: no need strdup - can pass name directly

Comment (yosefe) [reply src/uct/rocm/ipc/rocm_ipc_cache.c:220]: "rocm_ipc_cache_name"

File: src/ucs/config/parser.c (hunk 1 above)
Comment (hoopoepg) [reply src/ucs/config/parser.c:369]: technically yes, but value may come from external configuration and may contain few MB of data (@brminich seen issue where environment variable was filled by debug output) and process may crash due to stack overflow

File: src/ucs/config/parser.c (hunk 2 above)
Comment (hoopoepg) [reply src/ucs/config/parser.c:696]: alloca on processing of external data may cause stack overflow. do you think it is reasonable to use here?

File: src/ucs/config/parser.c (hunk 3 above)
Comment (hoopoepg) [reply src/ucs/config/parser.c]: renamed

File: src/ucs/stats/client_server.c (hunk 4 above)
Comment (hoopoepg) [reply src/ucs/stats/client_server.c]: renamed

File: src/ucs/stats/client_server.c (hunk 5 above)
Comment (hoopoepg) [reply src/ucs/stats/client_server.c]: renamed

File: src/ucs/stats/client_server.c (hunk 6 above)
Comment (hoopoepg) [reply src/ucs/stats/client_server.c]: renamed

File: src/ucs/stats/client_server.c (hunk 7 above)
Comment (hoopoepg) [reply src/ucs/stats/client_server.c]: renamed

File: src/uct/rocm/ipc/rocm_ipc_cache.c (hunk 8 above)
Comment (hoopoepg) [reply src/uct/rocm/ipc/rocm_ipc_cache.c]: > no need strdup - can pass name directly

will not work:
https://github.com/openucx/ucx/blob/209aae62d5aff3b4b5da07697c5240cdb9985990/src/uct/rocm/ipc/rocm_ipc_ep.c#L21
name is on-stack variable

Comment (hoopoepg) [reply src/uct/rocm/ipc/rocm_ipc_cache.c]: > "rocm_ipc_cache_name"

renamed

File: src/ucs/config/parser.c (hunk 2 above)
Comment (yosefe) [reply src/ucs/config/parser.c:696]: ok
so can you pls shorter allocation names for ucs_strdup?
for example, config_range_spec_str

Comment (hoopoepg) [reply src/ucs/config/parser.c]: renamed

File: src/ucs/config/parser.c [hunk 9]
```diff
@@ -957,7 +957,7 @@ int ucs_config_sscanf_table(const char *buf, void *dest, const void *arg)
     char *name, *value, *saveptr2;
     ucs_status_t status;
 
-    tokens = strdup(buf);
+    tokens = ucs_strdup(buf, "config_sscanf_table tokens");
```
Comment (yosefe) [inline src/ucs/config/parser.c:960]: config_sscanf_table

Comment (hoopoepg) [reply src/ucs/config/parser.c]: renamed

**Review by yosefe — APPROVED** [review]

**Review by dmitrygx — APPROVED** [review]

File: src/ucs/config/parser.c [hunk 10]
```diff
@@ -760,7 +760,7 @@ int ucs_config_sscanf_array(const char *buf, void *dest, const void *arg)
     int ret;
     unsigned i;
 
-    str_dup = strdup(buf);
+    str_dup = ucs_strdup(buf, "config_scanf_array str_dup");
```
Comment (brminich) [inline src/ucs/config/parser.c:763]: it is inconsistent with other changes made in this PR. You did not add `str_dup` to the name of other `ucs_strdup` calls

Comment (hoopoepg) [reply src/ucs/config/parser.c:763]: not sure I got your message... str_dup is variable name

Comment (brminich) [reply src/ucs/config/parser.c:763]: i mean that the name passed to `ucs_strdup` contains `str_dup`. I'd use `config_scanf_array`

Comment (hoopoepg) [reply src/ucs/config/parser.c]: ok, renamed

**Review by brminich — APPROVED** [review]

**Review by yosefe — APPROVED** [review]

**Pull request merged**